fairmq/MemoryResourceTools.h

```diff
@@ -0,0 +1,117 @@
+#include <fairmq/MemoryResources.h>
+
+namespace fair {
+namespace mq {
+
+using ByteSpectatorAllocator = SpectatorAllocator<fair::mq::byte>;
+using BytePmrAllocator = boost::container::pmr::polymorphic_allocator<fair::mq::byte>;
+
+//_________________________________________________________________________________________________
+// return the message associated with the container or nullptr if it does not
+// make sense (e.g. when
+// we are just watching an existing message or when the container is not using
+// FairMQMemoryResource
+// as backend).
+template<typename ContainerT>
+// typename std::enable_if<
+//    std::is_base_of<
+//        boost::container::pmr::polymorphic_allocator<typename
+//        ContainerT::value_type>,
+//        typename ContainerT::allocator_type>::value == true,
+//    FairMQMessagePtr>::type
+FairMQMessagePtr getMessage(ContainerT &&container_, FairMQMemoryResource *targetResource = nullptr)
+{
+    auto container = std::move(container_);
+    auto alloc = container.get_allocator();
+
+    auto resource = dynamic_cast<FairMQMemoryResource *>(alloc.resource());
+    if (!resource && !targetResource) {
+        throw std::runtime_error("Neither the container or target resource specified");
+    }
+    size_t containerSizeBytes = container.size() * sizeof(typename ContainerT::value_type);
+    if ((!targetResource && resource)
+        || (resource && targetResource && resource->is_equal(*targetResource))) {
+        auto message = resource->getMessage(static_cast<void *>(
+            const_cast<typename std::remove_const<typename ContainerT::value_type>::type *>(
+                container.data())));
+        if (message)
+            message->SetUsedSize(containerSizeBytes);
+        return std::move(message);
+    } else {
+        auto message = targetResource->getTransportFactory()->CreateMessage(containerSizeBytes);
+        std::memcpy(static_cast<fair::mq::byte *>(message->GetData()),
+                    container.data(),
+                    containerSizeBytes);
+        return std::move(message);
+    }
+};
+
+//_________________________________________________________________________________________________
+/// Return a vector of const ElemT, no wonership transfer.
+/// Resource must be kept alive throughout the lifetime of the
+/// container and associated message.
+template<typename ElemT>
+std::vector<const ElemT, boost::container::pmr::polymorphic_allocator<const ElemT>> getVector(
+    size_t nelem,
+    SpectatorMessageResource *resource)
+{
+    return std::vector<const ElemT, SpectatorAllocator<const ElemT>>(
+        nelem, SpectatorAllocator<ElemT>(resource));
+};
+
+//_________________________________________________________________________________________________
+/// Return a vector of const ElemT, takes ownership of the message
+template<typename ElemT>
+std::vector<const ElemT, OwningMessageSpectatorAllocator<const ElemT>>
+    getVector(size_t nelem, FairMQMessagePtr message)
```

Comment:
OK, so this is the intended interface. I'd vote strongly against inclusion of it because:
* The return type name is very long and I expect nobody to be able to spell it out. `auto` can help, but not so much when you want to store it as a data member.
* Users still see `std::vector` in the type and expect it to behave like a vector. I.e. `push_back`, `reserve`, and other operations that reallocate must work correctly. IIUC that's not the case.
* A simple new type with `begin`, `end`, `operator[]`, and `size()` that takes ownership of the `message` and provides an STL interface to read-only access seems easier to create than the current solution (`begin` and `end` could simply return a pointer, no need for a full-blown iterator type) and involves no UB.

Reply:
yes, I agree that custom containers that would allow direct adoption would be ideal, standard containers are just not designed to do this. One could try to make it work better (I think it would be possible to get the full expected functionality), but given the fact that gcc8 does not like this (libstdc++ static_asserts that T is not cv) this specific implementation has to go (or grow) anyway. I don't quite understand the need for this assertion as some functionality simply works with const elements, but they decided to start enforcing the requirement for T to be erasable, maybe for consistency.
I still think this is a useful exercise as I like the idea of containers that take ownership of the message together with the data so no additional book keeping is needed along the data path. I'll definitely give it more thought :)
- about the long type names - if we are to use allocators, the type names will be long anyway:)

Reply:
> I don't quite understand the need for this assertion as **some** functionality simply works with const elements

You already answered your question. ;-)

> if we are to use allocators, the type names will be long anyway

Unless we start using `std::pmr::vector` consistently.